### PR TITLE
syntax: Extra diagnostics for `_` used in an identifier position

### PR DESCRIPTION
**Pull request by petrochenkov**
Closes https://github.com/rust-lang/rust/issues/32501

### CODE REVIEW

File: src/libsyntax/parse/parser.rs [hunk 1]
```diff
@@ -572,9 +572,13 @@ impl<'a> Parser<'a> {
                 self.bug("ident interpolation not converted to real token");
             }
             _ => {
-                let token_str = self.this_token_to_string();
-                Err(self.fatal(&format!("expected ident, found `{}`",
-                                    token_str)))
+                let mut err = self.fatal(&format!("expected identifier, found `{}`",
+                                                  self.this_token_to_string()));
+                if self.token == token::Underscore {
+                    err.fileline_note(self.span, "`_` is not an identifier but a wildcard pattern \
```
Comment (nodakai) [inline src/libsyntax/parse/parser.rs]: According to the reference, a "wildcard pattern" is `..` (two dots) which can match with multiple data fields, as in
```rust
struct S(i32, i32);
if let S(..) = S(0, 1);
```
The underscore seems to be called a "placeholder".

File: src/libsyntax/parse/parser.rs [hunk 2]
```diff
@@ -572,9 +572,13 @@ impl<'a> Parser<'a> {
...
+                                                  and patterns are not expected in this position");
```
Comment (rkruppe) [inline src/libsyntax/parse/parser.rs]: I think this second half of the message could be omitted. It's a bit awkward to repeat what's expected here. The important piece of information that many users are missing is that `_` is not an identifier. "expected identifier" covers that it's not legal in this position.

**Comment by rkruppe** [comment]
:+1: while the referenced issue is a bit artificial, people are regularly confused by `_` being special. Anything which highlights that it's not an identifier will help with that.

File: src/libsyntax/parse/parser.rs (hunk 2 above)
Comment (petrochenkov) [inline src/libsyntax/parse/parser.rs]: Yeah, probably.

File: src/libsyntax/parse/parser.rs (hunk 1 above)
Comment (petrochenkov) [inline src/libsyntax/parse/parser.rs]: I don't know, the compiler uses "wildcard", [wikipedia](https://en.wikipedia.org/wiki/Pattern_matching) uses it too and that part of the reference looks ancient and not well-maintained.

**Comment by Manishearth** [comment]
@bors r-

ugh, it picked up the r=manishearth from my comment

**Pull request merged**